FeedStoreChallenge/Infrastructure/Core Data/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,86 @@
+//
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import CoreData
+
+public final class CoreDataFeedStore: FeedStore {
+	private static let modelName = "FeedStore"
+	private static let model = NSManagedObjectModel(name: modelName, in: Bundle(for: CoreDataFeedStore.self))
+
+	private let container: NSPersistentContainer
+	private let context: NSManagedObjectContext
+
+	struct ModelNotFound: Error {
+		let modelName: String
+	}
+
+	public init(storeURL: URL) throws {
+		guard let model = CoreDataFeedStore.model else {
+			throw ModelNotFound(modelName: CoreDataFeedStore.modelName)
+		}
+
+		container = try NSPersistentContainer.load(
+			name: CoreDataFeedStore.modelName,
+			model: model,
+			url: storeURL
+		)
+		context = container.newBackgroundContext()
+	}
+
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		perform { context in
+			do {
+				if let cache = try ManagedCache.find() {
+					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
+	}
+
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		perform { context in
+			do {
+				try CoreDataFeedStore.deleteCacheFeedIfNeeded(context: context)
+
+				let cache = ManagedCache(context: context)
+				cache.feed = ManagedFeedImage.images(from: feed, in: context)
+				cache.timestamp = timestamp
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
+	}
+
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		perform { context in
+			do {
+				try CoreDataFeedStore.deleteCacheFeedIfNeeded(context: context)
+				try context.save()
```

Comment:
In the case where there was no previously stored cache, `context.delete` will not be called, thus there will be no changes in the context and no need to save it. Otherwise, Core Data may perform unnecessary work.
It's possible to use `context.hasChanges` to check if saving is needed.
```suggestion
				if context.hasChanges {
					try context.save()
				}
```